hazelcast/src/main/java/com/hazelcast/jet/impl/connector/StreamEventJournalP.java

```diff
@@ -386,24 +396,41 @@ public int preferredLocalParallelism() {
 
         @Override
         public void init(@Nonnull Context context) {
-            if (clientXml != null) {
-                initRemote();
+            // The order is important.
+            // If dataLinkConfig is specified prefer it to clientXml
+            if (dataLinkRef != null) {
+                HzClientDataLinkFactory hzClientDataLinkFactory = getDataStoreFactory(context.hazelcastInstance(),
+                        dataLinkRef.getName());
+                HazelcastInstance client = hzClientDataLinkFactory.getDataLink();
+                findRemotePartitionCount(client);
+            } else if (clientXml != null) {
+                findRemotePartitionCountUsingNewClient();
             } else {
+                FunctionEx<? super HazelcastInstance, ? extends EventJournalReader<E>>
+                        eventJournalReaderSupplier = clusterMetaSupplierParams.getEventJournalReaderSupplier();
                 PermissionsUtil.checkPermission(eventJournalReaderSupplier, context);
                 initLocal(context.hazelcastInstance().getPartitionService().getPartitions());
             }
         }
 
-        private void initRemote() {
+        // Get remotePartitionCount from new HazelcastInstance
+        private void findRemotePartitionCountUsingNewClient() {
             HazelcastInstance client = newHazelcastClient(asClientConfig(clientXml));
             try {
-                HazelcastClientProxy clientProxy = (HazelcastClientProxy) client;
-                remotePartitionCount = clientProxy.client.getClientPartitionService().getPartitionCount();
+                findRemotePartitionCount(client);
             } finally {
                 client.shutdown();
             }
         }
 
+        // Get remotePartitionCount from given HazelcastInstance
+        private void findRemotePartitionCount(HazelcastInstance client) {
+            HazelcastClientProxy clientProxy = (HazelcastClientProxy) client;
+            ClientPartitionService clientPartitionService = clientProxy.client.getClientPartitionService();
+            // The implementation of getPartitionCount is using an AtomicInteger internally. So it is thread-safe
+            remotePartitionCount = clientPartitionService.getPartitionCount();
```

Comment:
Just a note: This call might throw `HazelcastClientOfflineException` if the client is configured with the async start. Can it cause any problems?

Reply:
I don't know. The exception was not getting caught since the beginning. 

As a general rule I can say that in case of any exception the processor does not start. So in the scenario that you have described it seems okay to me that the processor fails.